### ISSUE TITLE
Should we re-trigger the "When Touching" hat while it's running

### ISSUE THREAD

**Issue by BryceLTaylor** [LLK/scratch-gui#2301]
This came up during the smoke test.  I expected it to re-trigger, but I can see arguments for both ways.  @ericrosenbaum said maybe we should discuss it.

**Comment by griffpatch** [comment]
Well, when sprite clicked triggers only when the mouse is pressed, so I was
expecting this to work the same, however... The wording does imply it might
trigger (touching, rather than touches).

However, of the scripts linked to this hat block run over multiple game
cycles then it will constantly be restarted, the scripts will not run as
expected and much confusion would ensue I believe.

**Comment by kyleplo** [comment]
I think it shouldn't re-trigger, because other events blocks, such as `when key [ v] pressed` only trigger when the key is first pressed. To get it to happen constantly, users should use `<touching [ v]>`

**Comment by TheLogFather** [comment]
I think there are two senses in which "when touching" could be understood and, two behaviours for each – so making a total of *four* possible behaviours for such a hat block:

1) Every frame for which the touching is true, the script gets triggered, regardless of the current running status of the script.
2) Every frame for which the touching is true, the script gets triggered, *unless* it is currently running.
3) The script triggers on first frame that the touching becomes true, regardless of its current running status, and it doesn't run again while touching remains true (so can only trigger again after touching has switched back to false).
4) The script triggers on first frame that the touching becomes true, *unless* it is currently running, and it doesn't run again as long as touching remains true.

The first two behaviours come from reading it more as "*while* touching".

It seems to me that the only current (i.e. in Scratch 2) hat block that could possibly have the same choice of behaviours would be "when timer".

All other hat blocks are responding to discrete events –even "when key pressed" is responding to discrete 'key down' and 'key repeat' OS events, so it doesn't have the sense of 'continuousness' that the "key pressed" boolean has (though it's quite possible most Scratchers tend not to think of it like that...)

I would therefore argue that the *only* existing hat block to be potentially comparable to "when touching" is "when timer" – and that effectively has behaviour four (i.e. "when timer>N" has the sense of "do this at the point timer becomes > N", rather than "do this *while* timer > N", and it doesn't re-trigger while running).

----

Since behaviours 1 & 2 would be *completely new territory* for Scratch hat blocks, I'm assuming that the behaviour being discussed here is between 3 and 4...?

Either way, I would concur with griffpatch above in saying that it's better to *not* trigger while the script is currently running.

I can imagine it would be fairly typical for a Scratcher to begin a *sequence* of changes when two sprites begin touching each other (and both sprites may then have changes once they detect touching – though I can easily see all sorts of race conditions ahead as the Scratcher tries to use *two* "when touching" hat blocks, one in each sprite...)

I would therefore concur with griffpatch above that I'd expect fewer problems if it took behaviour four (i.e. cannot re-trigger if it is currently running).

----

Just to give an overview of current hat block behaviours in Scratch 2...

The hat blocks which are allowed to re-trigger while running are "when I receive", "when backdrop switches" and "when mouse clicked".

That leaves "when key pressed" and "when timer", which will *not* re-trigger while running.

**Comment by Wowfunhappy** [comment]
The current language of this block (when it's enabled in the block palette) is`when this sprite touches`.

This language (correctly) implies that the block is edge-activated. It doesn't say "when touching", it's "when _touches_". This is _perfect_ wording and it has worked well with my students.

So the question is whether the event should restart existing threads. My gut feeling is that events which are edge-activated should _not_ restart existing threads, events should be either one or the other but not both.

tl;dr I don't think you should change anything here. This block is great and has been extremely successful with my students.
